node/src/components/deploy_buffer.rs

```diff
@@ -422,6 +439,80 @@ impl Drop for DeployBufferMetrics {
     }
 }
 
+mod prune {
+    use super::*;
+
+    /// Prunes a deploy from an individual DeployCollection, returns the total
+    /// deploys pruned
+    pub fn prune_deploys(deploys: &mut DeployCollection, expired: DeployHash) -> usize {
+        let initial_len = deploys.len();
+        deploys.retain(|k, v| k != &expired && !v.dependencies().contains(k));
+        initial_len - deploys.len()
+    }
```

Comment:
We should prune recursively although I have no idea how complex (in CPU terms) it will end up :grimacing: I don't think we have another choice though – if we don't prune right away, those deploys will never be removed.